paasta_tools/tron_tools.py

```diff
@@ -594,18 +597,35 @@ def get_node_selectors(self) -> Dict[str, str]:
     def get_node_affinities(self) -> Optional[List[Dict[str, Union[str, List[str]]]]]:
         """Converts deploy_whitelist and deploy_blacklist in node affinities.
 
-        note: At the time of writing, `kubectl describe` does not show affinities,
+        NOTE: At the time of writing, `kubectl describe` does not show affinities,
         only selectors. To see affinities, use `kubectl get pod -o json` instead.
+
+        WARNING: At the time of writing, we only used requiredDuringSchedulingIgnoredDuringExecution node affinities in Tron as we currently have
+        no use case for preferredDuringSchedulingIgnoredDuringExecution node affinities.
         """
         requirements = allowlist_denylist_to_requirements(
             allowlist=self.get_deploy_whitelist(),
             denylist=self.get_deploy_blacklist(),
         )
+        node_selectors = self.config_dict.get("node_selectors", {})
         requirements.extend(
             raw_selectors_to_requirements(
-                raw_selectors=self.config_dict.get("node_selectors", {}),  # type: ignore
+                raw_selectors=self.config_dict.get("node_selectors", {}),
             )
         )
+
+        # PAASTA-18198: To improve AZ balance with Karpenter, we temporarily allow specifying zone affinities per pool
+        pool_node_affinities = load_system_paasta_config().get_pool_node_affinities()
+        if pool_node_affinities and self.get_pool() in pool_node_affinities:
+            current_pool_node_affinities = pool_node_affinities[self.get_pool()]
```

Comment:
what does this dictionary returns?

Reply:
this is a puppet-defined dict (see `paasta_tools::public_config::pool_node_affinities` internally) that defines a default set of affinities for certain pools (speaking of which, we should probably add an entry for the tron pool :p):
```
  default:
    topology.kubernetes.io/zone:
      - us-west-2a
      - us-west-2b
```

Reply:
ahh i see it returns the azs affinities:
```
paasta_tools::public_config::pool_node_affinities:
  # pnw-prod default should only run on us-west-2a and us-west-2b (unless there's a deploy_whitelist override)
  default:
    topology.kubernetes.io/zone:
      - us-west-2a
      - us-west-2b
 ```

Reply:
ahh sorry just saw your answer, yea makes sense